timelock-agent/src/main/java/com/palantir/atlasdb/timelock/paxos/LeadershipContextFactory.java

```diff
@@ -0,0 +1,165 @@
+    }
+
+    @Value.Derived
+    public LeaderPinger leaderPinger() {
+        return leaderPingerFactory().create(this);
+    }
+
+    @Value.Derived
+    LeaderPingHealthCheck leaderPingHealthCheck() {
+        return leaderPingHealthCheckFactory().create(this);
+    }
+
+    @Override
+    public LeadershipContext create(Client client) {
+        return ImmutableClientAwareComponents.builder()
+                .from(this)
+                .proxyClient(client)
+                .build()
+                .leadershipContext();
+    }
+
+    @Value.Derived
+    public AutobatchingLeadershipObserverFactory leadershipObserverFactory() {
+        return TimelockLeadershipMetrics.createFactory(metrics());
+    }
+
+    @Value.Immutable
+    abstract static class ClientAwareComponents implements
+            Dependencies.ClientAwareComponents,
+            Dependencies.LeadershipMetrics {
+
+        public abstract Client proxyClient();
+
+        @Value.Derived
+        Client paxosClient() {
+            return useCase().resolveClient(proxyClient());
+        }
+
+        @Value.Derived
+        PaxosLearner localLearner() {
+            return components().learner(paxosClient());
+        }
+
+        @Value.Derived
+        public PingableLeader localPingableLeader() {
+            return components().pingableLeader(paxosClient());
+        }
+
+        @Value.Derived
+        TimelockLeadershipMetrics leadershipMetrics() {
+            return ImmutableTimelockLeadershipMetrics.builder().from(this).build();
+        }
+
+        @Value.Derived
+        public PaxosLeadershipEventRecorder eventRecorder() {
+            return leadershipMetrics().eventRecorder();
+        }
+
+        @Value.Derived
+        LeaderElectionService leaderElectionService() {
+            return new LeaderElectionServiceBuilder()
+                    .leaderPinger(leaderPinger())
+                    .leaderUuid(leaderUuid())
+                    .pingRate(runtime().get().pingRate())
+                    .randomWaitBeforeProposingLeadership(runtime().get().maximumWaitBeforeProposingLeadership())
+                    .eventRecorder(eventRecorder())
+                    .knowledge(localLearner())
+                    .acceptorClient(networkClientFactories().acceptor().create(paxosClient()))
+                    .learnerClient(networkClientFactories().learner().create(paxosClient()))
+                    .decorateProposer(this::instrumentProposer)
+                    .build();
+        }
+
+        @Value.Derived
+        LeadershipContext leadershipContext() {
+            return ImmutableLeadershipContext.builder()
+                    .leadershipMetrics(leadershipMetrics())
+                    .leaderElectionService(leaderElectionService())
+                    .build();
+        }
+
+        private PaxosProposer instrumentProposer(PaxosProposer uninstrumentedPaxosProposer) {
+            return metrics().instrument(
```

Comment:
This doesn't work because this work hasn't been done for `TaggedMetricRegistry`s. There isn't a single class that's been instrumented with `MetricRegistry`.

Reply:
I can add the annotation to "future"-proof it, but considering it doesn't do anything and it's not related to this PR, I can do that in a later PR.